File: client/src/game_logic.rs
use crate::state::DiskColor;

// 方向を表す定数
const DIRECTIONS: [(i32, i32); 8] = [
    (-1, -1), (-1, 0), (-1, 1), 
    (0, -1),           (0, 1), 
    (1, -1),  (1, 0),  (1, 1)
];

// ボードの大きさ
const BOARD_SIZE: usize = 8;

// 有効な手かどうかをチェックする関数
pub fn is_valid_move(board: &Vec<Vec<Option<DiskColor>>>, x: usize, y: usize, color: &DiskColor) -> bool {
    // すでに石が置かれていたら無効
    if board[y][x].is_some() {
        return false;
    }
    
    // 8方向をチェック
    for (dx, dy) in &DIRECTIONS {
        let mut nx = x as i32 + dx;
        let mut ny = y as i32 + dy;
        
        // 盤外なら次の方向へ
        if nx < 0 || nx >= BOARD_SIZE as i32 || ny < 0 || ny >= BOARD_SIZE as i32 {
            continue;
        }
        
        // 隣が相手の色でなければ次の方向へ
        let opposite_color = match color {
            DiskColor::Black => DiskColor::White,
            DiskColor::White => DiskColor::Black,
        };
        
        match board[ny as usize][nx as usize] {
            Some(ref c) if c == &opposite_color => {},
            _ => continue,
        }
        
        // さらに進む
        nx += dx;
        ny += dy;
        
        // 盤外に出るまで進む
        while nx >= 0 && nx < BOARD_SIZE as i32 && ny >= 0 && ny < BOARD_SIZE as i32 {
            match board[ny as usize][nx as usize] {
                // 自分の色を見つけたら有効
                Some(ref c) if c == color => return true,
                // 相手の色なら進み続ける
                Some(ref c) if c == &opposite_color => {
                    nx += dx;
                    ny += dy;
                },
                // 空白なら無効（連続していない）
                _ => break,
            }
        }
    }
    
    false
}

// 有効な手を計算する関数
pub fn calculate_valid_moves(board: &Vec<Vec<Option<DiskColor>>>, current_player: &DiskColor) -> Vec<(usize, usize)> {
    let mut valid_moves = Vec::new();
    
    for y in 0..BOARD_SIZE {
        for x in 0..BOARD_SIZE {
            if is_valid_move(board, x, y, current_player) {
                valid_moves.push((x, y));
            }
        }
    }
    
    valid_moves
}

File: client/utils/gameLogic.ts
import { DiskColor } from "@/store/gameStore";

const DIRECTIONS = [
  [-1, -1],
  [-1, 0],
  [-1, 1],
  [0, -1],
  [0, 1],
  [1, -1],
  [1, 0],
  [1, 1],
];

const BOARD_SIZE = 8;

export const calculateValidMoves = (
  board: Array<Array<DiskColor | null>>,
  currentPlayer: DiskColor
): Array<[number, number]> => {
  const validMoves: Array<[number, number]> = [];

  for (let y = 0; y < BOARD_SIZE; y++) {
    for (let x = 0; x < BOARD_SIZE; x++) {
      if (isValidMove(board, x, y, currentPlayer)) {
        validMoves.push([x, y]);
      }
    }
  }

  return validMoves;
};

export const isValidMove = (
  board: Array<Array<DiskColor | null>>,
  x: number,
  y: number,
  color: DiskColor
): boolean => {
  // セルが既に埋まっている場合は無効
  if (board[y][x] !== null) {
    return false;
  }

  const oppositeColor = color === "black" ? "white" : "black";

  // 8方向をチェック
  for (const [dx, dy] of DIRECTIONS) {
    let nx = x + dx;
    let ny = y + dy;

    // 盤外または相手の石がない場合はこの方向をスキップ
    if (
      nx < 0 ||
      nx >= BOARD_SIZE ||
      ny < 0 ||
      ny >= BOARD_SIZE ||
      board[ny][nx] !== oppositeColor
    ) {
      continue;
    }

    nx += dx;
    ny += dy;

    // この方向に進み、自分の石があるかチェック
    while (nx >= 0 && nx < BOARD_SIZE && ny >= 0 && ny < BOARD_SIZE) {
      if (board[ny][nx] === null) {
        // 空のセルがあれば無効
        break;
      }

      if (board[ny][nx] === color) {
        // 自分の石があれば有効な手
        return true;
      }

      // 相手の石の場合は進み続ける
      nx += dx;
      ny += dy;
    }
  }

  return false;
};
